supervisor/main.go

```diff
@@ -0,0 +1,183 @@
+	// Addr - address to start the HTTP server(host:port).
+	Addr string
+}
+
+// parseArgs returns the parsed arguments.
+func parseArgs() *args {
+	var args args
+	flag.StringVar(&args.CfgPath, "cfg", "cfg.json",
+		"path to Supervisor config.")
+	flag.StringVar(&args.Addr, "addr", "127.0.0.1:8080",
+		"address to start the HTTP server(host:port).")
+	flag.Parse()
+
+	return &args
+}
+
+// parseCfg parse the Supervisor JSON config.
+func parseCfg(path string) (*core.Cfg, error) {
+	// Check is the file exists.
+	if _, err := os.Stat(path); err != nil {
+		return nil, err
+	}
+
+	// Open the file.
+	jsonFile, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer jsonFile.Close()
+
+	// Set defaults.
+	cfg := core.Cfg{
+		InstancesDir: "/etc/tarantool/supervisor/instances",
+		TermTimeout:  30,
+	}
+
+	// Read and parse config.
+	byteValue, err := ioutil.ReadAll(jsonFile)
+	if err != nil {
+		return nil, err
+	}
+	if err = json.Unmarshal(byteValue, &cfg); err != nil {
+		return nil, err
+	}
+
+	// In the config, the time is indicated in seconds. Convert the value.
+	cfg.TermTimeout = cfg.TermTimeout * time.Second
+
+	return &cfg, nil
+}
+
+// handleZombie handles zombie child processes, if any, in a non-blocking style.
+func handleZombie(sv *core.Supervisor) {
+	// Get PID of the terminated Instance.
+	pid, err := syscall.Wait4(-1, nil, syscall.WNOHANG, nil)
+
+	// If the error "no child processes" (syscall.Errno(10)) or pid == 0 -
+	// this means that the process was stopped as planned, and the wait ()
+	// function has already been called.
+	if pid == 0 || (err != nil && errors.Is(err, syscall.Errno(10))) {
+		return
+	} else if err != nil {
+		log.Printf(`Can't get the PID of a process by using "Wait4". Error: "%v"`,
+			err)
+		return
+	}
+
+	// If the Restartable Instance flag is true, try to restart the Instance.
+	if id, err := sv.RestartAfterTermInstance(pid); err != nil {
+		log.Printf(`Can't restart the Instance. Old PID : %v. ID: %v. Error: "%v"`,
+			pid, id, err)
+	} else {
+		log.Printf("The Instance has been restarted. ID: %v", id)
+	}
+}
+
+// terminateGracefully terminates the service correctly.
+func terminateGracefully(sv *core.Supervisor, srv *http.Server,
+	timeout time.Duration, done chan bool) {
+	// First of all, shut down the HTTP server to avoid reciving a new request.
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	if err := srv.Shutdown(ctx); err != nil {
+		log.Printf(`HTTP server shutdown error: "%v"`, err)
+	}
+	cancel()
+
+	// And now stop all running Instances.
+	sv.StopAllInstance()
+
+	done <- true
+}
+
+// startSignalHandling adds "SIGTERM" and "SIGINT" signal handling to
+// terminate gracefully and "SIGCHLD" to restart Instances.
+func startSignalHandling(sv *core.Supervisor, srv *http.Server,
+	termTimeout time.Duration, done chan bool) {
+	// The handling of the "SIGTERM" and "SIGINT" signals
+	// will be used to terminate gracefully.
+	sigTerm := make(chan os.Signal, 1)
```

Comment:
`sigTerm` -> `sigTermChan`.

Looks weird when you read from sigterm =)

Reply:
OK